File: server/src/main/java/bgu/spl/net/srv/Frame.java
package bgu.spl.net.srv;

import java.util.HashMap;
import java.util.Map;

public class Frame {
    private String command;
    private Map<String, String> headers;
    private String body;

    public Frame(String command, Map<String, String> headers, String body) {
        this.command = command;
        this.headers = headers;
        this.body = body;
    }

    public Frame(){
        this.command = null;
        this.headers = new HashMap<String, String>();
        this.body = null;
    }

    public String getCommand() {
        return command;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    public String getBody() {
        return body;
    }

    public String getKeyByHeader(String header) {
        return headers.get(header);
    }

    public void setCommand(String command) {
        this.command = command;
    }

    public void setHeaders(Map<String, String> headers) {
        this.headers = headers;
    }

    public void setBody(String body) {
        this.body = body;
    }

    public void addHeader(String key, String value) {
        headers.put(key, value);
    }

    public String toString() {
        String result = this.getCommand() + "\n";
        for (String key : this.getHeaders().keySet()) {
            result += key + ":" + this.getHeaders().get(key) + "\n";
        }
        result += "\n" + this.getBody() + "\u0000";
        return result;
    }
}

File: client/src/Frame.cpp
#include "Frame.h"

Frame::Frame(const string &command) : command(command), headers(), body() {}

Frame::~Frame() = default;

Frame::Frame(const Frame &other) : command(other.command), headers(other.headers), body(other.body) {}

Frame::Frame(Frame &&other) noexcept : command(std::move(other.command)), headers(std::move(other.headers)), body(std::move(other.body)) {}

Frame &Frame::operator=(const Frame &other)
{
    if (this != &other)
    {
        command = other.command;
        headers = other.headers;
        body = other.body;
    }
    return *this;
}

Frame &Frame::operator=(Frame &&other) noexcept
{
    if (this != &other)
    {
        command = std::move(other.command);
        headers = std::move(other.headers);
        body = std::move(other.body);
    }
    return *this;
}
void Frame::addHeader(const string &key, const string &value)
{
    headers[key] = value;
}

void Frame::setBody(const string &body)
{
    this->body = body;
}

string Frame::getCommand() const
{
    return command;
}

string Frame::getHeader(const string &key) const
{
    auto it = headers.find(key);
    return (it != headers.end()) ? it->second : "";
}

string Frame::getBody() const
{
    return body;
}

Frame Frame::parseFrame(const string &frameString)
{
    vector<string> lines = splitIntoLines(frameString);

    Frame frame(lines[0]);
    for (int i = 1; i < lines.size() - 1; i++)
    {
        int colonPos = lines[i].find(':');
        string key = lines[i].substr(0, colonPos);
        string value = lines[i].substr(colonPos + 1);
        frame.addHeader(key, value);
    }

    frame.setBody(lines[lines.size() - 1]);
    return frame;
}

vector<string> Frame::splitIntoLines(const string &input)
{
    vector<string> lines;
    istringstream stream(input);
    string line;

    while (getline(stream, line))
    {
        lines.push_back(line);
    }

    return lines;
}

string Frame::toString() const
{
    string result = command + "\n";
    for (const auto &header : headers)
    {
        result += header.first + ": " + header.second + "\n";
    }
    if (!body.empty())
    {
        result += "\n" + body + "\n";
    }
    result += '\0';
    return result;
}